File: 0394-decode-string/0394-decode-string.cpp
class Solution {
public:
    string decodeString(string s) {
        stack<char> st;
        for(int i = 0; i<s.size();i++)
        {
            if(s[i] ==']')
            {
                string decodedString="";
                while(!st.empty() && st.top()!='[')
                {
                    decodedString += st.top();
                    st.pop();
                }
                
                st.pop();
                int num = 0;
                int base = 1;
                while(!st.empty() && isdigit(st.top()))
                {
                    num = num + (st.top() - '0')*base;
                    base*=10;
                    st.pop();
                }
                while(num!=0)
                {
                    for(int i = decodedString.size()-1; i>=0;i--)
                    {
                        st.push(decodedString[i]);
                    }
                    num--;
                }
            }
            else
            {
                st.push(s[i]);
            }
        }
        string ans = "";
       while(!st.empty())
       {
        ans = st.top()+ans;
        st.pop();
       }
       return ans;
    }
};

File: 0394-decode-string/0394-decode-string.java

class Solution {
    public String decodeString(String s) {
        Stack<String> stack = new Stack<>();  // Stack to store string components
        StringBuilder currentString = new StringBuilder();
        int k = 0;

        for (char ch : s.toCharArray()) {
            if (Character.isDigit(ch)) {
                k = k * 10 + (ch - '0');  // Build the full number for repetition
            } else if (ch == '[') {
                // Push current string and k onto stack
                stack.push(currentString.toString());  
                stack.push(String.valueOf(k));
                currentString = new StringBuilder(); // Reset for new substring
                k = 0;  // Reset repeat count
            } else if (ch == ']') {
                int repeatCount = Integer.parseInt(stack.pop());  // Get stored repeat count
                String previousString = stack.pop();  // Get previous string from stack
                String repeatedString = currentString.toString().repeat(repeatCount);
                currentString = new StringBuilder(previousString + repeatedString);  // Append expanded substring
            } else {
                currentString.append(ch);  // Build substring
            }
        }
        return currentString.toString();
    }
}
